tests/dialer_test.go

```diff
@@ -0,0 +1,147 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/certs"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/stdlib"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sqladmin/v1"
+)
+
+func TestClientHandlesSSLReset(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping dialer integration tests")
+	}
+	newClient := func(c *http.Client) *proxy.Client {
+		return &proxy.Client{
+			Port: 3307,
+			Certs: certs.NewCertSourceOpts(c, certs.RemoteOpts{
+				UserAgent:      "cloud_sql_proxy/test_build",
+				IPAddrTypeOpts: []string{"PUBLIC", "PRIVATE"},
+			}),
+			Conns: proxy.NewConnSet(),
+		}
+	}
+	connectToDB := func(c *proxy.Client) (*sql.DB, error) {
+		var (
+			dbUser = *postgresUser
+			dbPwd  = *postgresPass
+			dbName = *postgresDb
+		)
+		dsn := fmt.Sprintf("user=%s password=%s database=%s", dbUser, dbPwd, dbName)
+		config, err := pgx.ParseConfig(dsn)
+		if err != nil {
+			return nil, err
+		}
+		config.DialFunc = func(ctx context.Context, network, instance string) (net.Conn, error) {
+			return c.DialContext(ctx, *postgresConnName)
+		}
+		dbURI := stdlib.RegisterConnConfig(config)
+		return sql.Open("pgx", dbURI)
+	}
+	resetSSL := func(c *http.Client) error {
+		svc, err := sqladmin.NewService(context.Background(), option.WithHTTPClient(c))
+		if err != nil {
+			return err
+		}
+		project, _, instance, _, _ := proxy.ParseInstanceConnectionName(*postgresConnName)
+		t.Log("Reseting SSL config.")
+		op, err := svc.Instances.ResetSslConfig(project, instance).Do()
+		if err != nil {
+			return err
+		}
+		for {
+			t.Log("Waiting for operation to complete.")
+			op, err = svc.Operations.Get(project, op.Name).Do()
+			if err != nil {
+				return err
+			}
+			if op.Status == "DONE" {
+				t.Log("reset SSL config operation complete")
+				break
+			}
+			time.Sleep(time.Second)
+		}
+		return nil
+	}
+
+	// SETUP: create HTTP client and proxy client, then connect to database
+	src, err := google.DefaultTokenSource(context.Background(), proxy.SQLScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client := oauth2.NewClient(context.Background(), src)
+	proxyClient := newClient(client)
+
+	db, err := connectToDB(proxyClient)
+	if err != nil {
+		t.Fatalf("failed to connect to DB: %v", err)
+	}
+
+	// Begin database transaction
+	tx, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tx.Rollback()
+
+	resetSSL(client)
+
+	// Re-dial twice, once to invalidate config, once to establish connection
+	var attempts int
+	for {
+		t.Log("Re-dialing instance")
+		_, err = proxyClient.DialContext(context.Background(), os.Getenv("POSTGRES_CONNECTION_NAME"))
```

Comment:
Should this env var be a reference to an existing var? I think you can pass these in via flags as well

Reply:
Yes, that's a good point.